### PR TITLE
Make groupsOf and greedyGroupsOf fn families tail-recursive

### LINKED ISSUE #75: Group methods will crash if used with a large list
Several methods are implemented in a way that is not tail recursive. I'm not sure what's the current state in elm and in js regarding tail call optimization (see https://lisplog.org/elm_converts_tail_calls_to_loops.html) or if [elm-lang/trampoline](http://package.elm-lang.org/packages/elm-lang/trampoline/1.0.1/Trampoline) could help. Open for discussion :)

#### Reproduce issue

```elm
list : List (List Int)
list =
    List.range 0 100000
    |> List.Extra.group
```
image

https://ellie-app.com/3MxzbGDZdNSa1/1

#### Related issues
#61

### PR DESCRIPTION
**Pull request by Janiczek**
Closes #75

I've hit this during Advent of Code 😭 😄

### CODE REVIEW

File: src/List/Extra.elm [hunk 1]
```diff
@@ -1863,25 +1863,34 @@ skipped and not appear in any groups.
...
-    else
-        []
+                    okayLength =
+                        size == List.length thisGroup
+                in
+                if okayArgs && okayLength then
+                    go rest (thisGroup :: acc)
+
+                else
+                    go rest acc
```
Comment (jfmengels) [inline src/List/Extra.elm]: Can't we skip one iteration by calling `List.reverse acc` in this case?

File: src/List/Extra.elm [hunk 2]
```diff
@@ -1863,25 +1863,34 @@ skipped and not appear in any groups.
...
+                    rest =
+                        List.drop step xs
 
-        okayLength =
-            size == List.length thisGroup
-    in
-    if okayArgs && okayLength then
-        thisGroup :: groupsOfWithStep size step xs_
+                    okayArgs =
+                        size > 0 && step > 0
```
Comment (jfmengels) [inline src/List/Extra.elm]: This can probably be computed once in `groupsOfWithStep` but outside of `go`

File: src/List/Extra.elm [hunk 3]
```diff
@@ -1863,25 +1863,32 @@ skipped and not appear in any groups.
...
+                        thisGroup =
+                            List.take size xs
 
-        okayLength =
-            size == List.length thisGroup
-    in
-    if okayArgs && okayLength then
-        thisGroup :: groupsOfWithStep size step xs_
+                        rest =
+                            List.drop step xs
```
Comment (jfmengels) [inline src/List/Extra.elm]: You could move this definition to inside the if (then) condition. That way, we won't compute the rest if `size /= List.length thisGroup`.

Comment (Janiczek) [reply src/List/Extra.elm]: Done in fb47d4a, thanks!

**Comment by Chadtech** [comment]
Thanks for looking into this. Sorry for the delay, but I will check out this PR ASAP (later today or tomorrow).

**Comment by Chadtech** [comment]
I just benchmarked this here: https://ellie-app.com/gm9X8yfPLXMa1 (with a one small adjustment: I pulled `go` out into a top level function). There seems to be a performance drop primarily on small lists and small group sizes. 

(Ignore the fact that its labelled "isPermutationOf". Thats because I copy and pasted the "isPermutationOf" benchmark to get started).
Screen Shot 2022-01-08 at 02 56 25
Screen Shot 2022-01-08 at 02 56 37
Screen Shot 2022-01-08 at 02 56 52

Is it worth it? Probably. But what do you all think?

**Comment by Janiczek** [comment]
> Is it worth it?

I'd consider it worth it, as correctness issues IMHO should have more priority than performance issues

**Comment by jfmengels** [comment]
I agree with @Janiczek, correctness is more important than performance (at least as a generally available function like this). In a way, it's good that it's slower mostly for the small inputs because performance increases for those don't matter too much in the big picture anyway (I mean, you can still call this function 2 million times per second apparently!).

**Comment by Chadtech** [comment]
Sounds good to me. Thanks @Janiczek !

**Pull request merged**